### DISCUSSION TITLE
Security composable to use in pages

### DISCUSSION THREAD

**Discussion by maxdzin** [Baroshem/nuxt-security#217] [Ideas]
Hey @Baroshem, is it possible to set the security settings on the page directly?

It would be good to have since it requires to be set by certain conditions sometimes.
So, it would be nice to have some composable, similar to [useSeoMeta](https://nuxt.com/docs/api/composables/use-seo-meta).

And so in the `/pages/[slug].vue`:
```ts
// some logic to define if form with files upload is used
const canUploadFiles = computed<boolean>(() => true)

useSecurity({
  xssValidator: canUploadFiles.value ? false : {},
  corsHandler: {
    origin: '*',
    methods: canUploadFiles.value ? ['GET', 'HEAD'] : ['GET', 'HEAD', 'POST', 'PUT'],
    preflight: {
      statusCode: 204,
    }
  },
  // ... other route security settings
})
```

Whad do you  think?

**Comment by Baroshem** [comment]
Hey @maxdzin 

This is a really interesting feature!

I would completely see the usage of it as a similar approach to useSeoMeta. 

I am just not sure if its doable from the Nuxt side as right now Nuxt Security uses either runtime config or routeRules to configure server middlewares dynamically. 

@danielroe what do you think about it?

**Reply by danielroe** [reply]
There's already support for routeRules being defined directly in pages. Maybe this would be a good way to implement.

**Reply by maxdzin** [reply]
>There's already support for routeRules being defined directly in pages.

@danielroe Where can I see that?

**Reply by danielroe** [reply]
https://github.com/nuxt/nuxt/pull/20391

**Reply by Baroshem** [reply]
Wow this is amazing! I will try to create a PoC solution to see how it works :)

**Comment by Baroshem** [comment]
I tried to make this work but for some reason the composable I created does not work :(

```ts
import type {
  NuxtSecurityRouteRules
} from '../../types'

export function useSecurity (config: NuxtSecurityRouteRules & { headers?: Record<string, string> }) {
  const { headers, ...security } = config

  defineRouteRules({
    security,
    headers
  })
}
```

I tried adding normal `defineRouteRules` in the page and it works correctly. But it still needs to add this line manually in the nuxt config ts:

```ts
  experimental: {
    inlineRouteRules: true
  },
```

And at this point I wonder if it actually makes sense to create such composable.

Right now, users can easily configure per page routes like (after enabling it in the nuxt.config.ts):

```ts
defineRouteRules({
  headers: {
    'X-XSS-Protection': '1'
  },
  security: {
      rateLimiter: {
    tokensPerInterval: 3,
    interval: 60000,
  },
  }
})
```

Not sure if it makes sense to do that (Especially that it does not work)

**Reply by danielroe** [reply]
No, you can't use a composable with `defineRouteRules`. It is a macro and needs to be in the page where it is used.

**Reply by Baroshem** [reply]
That was fast :D 

So there is no way to implement such `useSecurity` composable that will change the per route configuration like this?

**Reply by maxdzin** [reply]
@Baroshem 
> And at this point I wonder if it actually makes sense to create such composable.
Right now, users can easily configure per page routes like (after enabling it in the nuxt.config.ts):

If that works then, yes, I think that there's no need to create a composable that just repeats such settings definition.
I'll try to check how it works soon.

**Reply by Baroshem** [reply]
The real question right now is whether to add it to the security documentation somewhere? How do you guys think?

**Reply by maxdzin** [reply]
It was a real problem for some projects and I suppose not only I who faced such a thing.
So, yes, that definitely must be in the docs!

**Reply by Baroshem** [reply]
Added in https://github.com/Baroshem/nuxt-security/pull/212/commits/724186f99c9b2710a075bb6ee6172b1334410438

**Reply by maxdzin** [reply]
@Baroshem when specifying like that
```ts
  defineRouteRules({
    security: {
      xssValidator: canUploadFiles.value ? false : {},
    },
  })
```
typecheck error appears:
```error TS2741: Property 'allowedMethodsRestricter' is missing in type '{ xssValidator: false | {}; }' but required in type 'NuxtSecurityRouteRules'.```

so I needed to specify the `allowedMethodsRestricter` for the security within those page settings:
```ts
  defineRouteRules({
    security: {
      xssValidator: canUploadFiles.value ? false : {},
      allowedMethodsRestricter: '*',
    },
  })
```

I think that could be possibly omitted if its default value is about to be used, right?

**Reply by Baroshem** [reply]
@maxdzin Yes, it fixed in 1.0.0-rc.1 :)

**Reply by maxdzin** [reply]
Perfect! Thank you!
